### DISCUSSION TITLE
Plans for bell character

### DISCUSSION THREAD

**Discussion by notcancername** [ghostty-org/ghostty#2710] [Feature Requests, Ideas] [answered]
Sup,
Ghostty currently doesn't seem to do anything when faced with a bell character (`printf \x7`).

https://github.com/ghostty-org/ghostty/blob/90c59f2462d4a2e43456b865646ac6b25d243380/src/termio/stream_handler.zig#L326-L327

What's planned for that?

**Comment by jcollie** [comment]
See #2231

**Comment by mitchellh** [comment]
The lack of a long term plan is part of the reason I haven't merged #2231. The other is that there are a lot of dependencies it brings in and I want to think about that a bit more. I think in general the plan for the bell character is to at the very least provide an option for (1) ignore (2) audio (3) visual (4) both. I'd love to hear any ideas people have in general. Once we have some solid idea of _what_ we want, we can bike shed on configuration styles and implementation choices.

**Reply by jcollie** [reply]
> The other is that there are a lot of dependencies it brings in and I want to think about that a bit more.

I'm not sure it really _adds_ dependencies. GStreamer is already a dependency of GTK, the changes in the Nix packages are mainly to make sure that the development packages and plugins are available in Nix development environments.

**Reply by rvdm** [reply]
Came up on this thread when searching for ways to configure behaviour when a bell gets triggered. One thing I like about iTerm (but not enough to switch back) is the MacOS icon badge it shows when a bell happens. 

I use some of my terminals for longer-running processes (irc 😄 ) and like it when I can see 'something happened' (for irc, when a private message arrives) without switching back to the terminal window. I typically silence the notifications, but in iTerm it notifies with something like 'terminal sounded a bell!'. 

For a basic 'parity with other terminals' feature set I would suggest to start with the 'ignore/audio/visual/notification' options. If these are seperately configurable there is sufficient room for change. An ultimate dream would be something like this, but I guess that's maybe over-engineering it:
```
bell-visual: true
bell-audio: false
bell-notification-message: "Ghostty wants your attention!" // disabled when unset, can be expanded with templated vars once they are available
bell-notification-script: "/usr/local/bin/ghosttynag.sh" // disabled when unset, can be expanded with templated vars once they are available
```

**Reply by ovesh** [reply]
> I think in general the plan for the bell character is to at the very least provide an option for (1) ignore (2) audio (3) visual (4) both.

Personally I like iTerm's icon badge and the bouncing icon.

**Reply by kfet** [reply]
What I find useful is icon badges and a notification. As others I tend to use a bel to indicate end of commands which I care to be aware of.

**Reply by vikanezrimaya** [reply]
I would rather have Ghostty wrap libcanberra or GSound (which is effectively a GObject wrapper around libcanberra) so it can use the system sound theme instead of something bespoke like the bundled bell sound in #2231 or a separate config option.

**Reply by cwarden** [reply]
I'm just looking for a visual bell that flashes the screen like xterm or alacritty.

**Reply by ftpd** [reply]
I'm seconding the idea of icon badge on macOS.

**Reply by dolmen** [reply]
A visual only flash in the terminal area (ex: temporary inverse foreground/background color, or apply a GL shader) could probably be implemented in a portable way (independant of the OS) without bringing huge dependencies (contrary to audio).

**Comment by jcollie** [comment]
In terms of configuration, I'd like to see a style similar to `shell-integration-features` where there's one option that can enable/disable many features. This takes advantage of generic code in Ghostty for handling these sorts of configs and it's a little more future-proof.

So you'd have something like the following: 

`bell-features = audio,no-visual,no-message`

For the initial PR, we can limit the features to audio bells only to minimize the areas of code being touched and make review easier.

**Comment by dolmen** [comment]
For the case of a visual representation of `BEL`, I think it would be helpful to have visual information of the location where the BEL occurred: the location of the cursor where the `BEL` has been printed. I'm thinking at flashing the whole terminal (using a GL shader?) or just the scroll region, but with a visual effect that highlight the source position of the cursor. 

I'm thinking at waves like when throwing a drop of water on a calm lake, seen from the sky. Multiple `BEL` can happen at different locations and that could to a rain-on-a-lake effect.

If a scroll event an asynchronous animation must continue to play still attached to the location where the `BEL` was printed, updated according to scrolling.

**Comment by dolmen** [comment]
About visual `BEL`, what do you think about:
* _synchronous_ (blocking terminal updates until animation is done)
* vs _asynchronous_ (animation starts when BEL is recieved, but lasts for a few milliseconds after flow continues). In that case, the display of multiple `BEL` happening has to be handled.

**Reply by jcollie** [reply]
IMHO blocking terminal updates would be a really bad idea. From what I know of the graphics rendering timeline it wouldn't be difficult to have a timestamp object that says "show the visual bell animation until the current time reaches this timestamp". When we receive a BEL character that timestamp gets set to `now + 500ms` (or whatever interval you like). That way the animation is shown for at most 500ms from the time the last BEL character was received.

**Reply by jpaugh** [reply]
Blocking terminal updates from a non-interactive script because of a UI update is really heavy-handed. I'd rather miss the second bel than have that.

Ignoring all bels during the animation would not be that bad, especially if the animation is short enough not to be annoying (Something like .2 seconds).

If you want to indicate every bel that comes in, just restart the animation. That will provide a visual "jolt" to the user, and it is reasonably intuitive to me (as a user).

Extending the animation would be ambiguous, and ultimately conveys the same meaning as ignoring the second bell. Even if the user understands why the animation lasts longer than usual, would a user be able to distinguish between 2 bels and 3?

**Comment by jneuhaus** [comment]
I'm in favor of an audible tone for BEL, although it does seem that other terminal applications do not send an audible (e.g. Terminator.) My use-case is monitoring a process while ssh into a system...I have a bash script which runs a process, checks the result, then sleeps (all in a loop.) If the process completes, I send a BEL (in another loop) which alerts me that the process is complete. I'm not in front of the system doing this, but nearby, which is why the audible is useful.

**Comment by pluiedev** [comment]
My 2¢ on this issue is that we should mostly rely on what the system defines to be an appropriate response to a generic system bell. On Linux, GTK has a [dedicated method for it](https://docs.gtk.org/gdk4/method.Surface.beep.html) and on macOS we have the lovely-named [`NSBeep`](https://developer.apple.com/documentation/appkit/nsbeep). A maximalist approach à la Kitty would cause very high maintenance burden for something that is IMO just an audiovisual flourish.

Of course, that isn't to say that we couldn't strike a balance between only offering only a system bell and the whole charcuterie plate of options Kitty provides, though each kind of bell needs to be considered individually:

- Audio: While relatively easy to implement on macOS, on Linux it's somewhat of a nightmare. [libcanberra is considered obsolete](https://discourse.gnome.org/t/is-libcanberra-deprecated/10336/4) and directly using GSound requires bundling our own sound theme in a way, like in #5326. This breaks the "everything looks and feels (and sounds) native" paradigm since we pretty much have to fall back to the default FreeDesktop sound theme, while on say KDE Plasma the rest of the system uses the Ocean sound theme, designed specifically for Plasma.
  
  Additionally, some systems (namely Plasma again) allows one to customize what sound should be played when a system bell is triggered. I think that's a much healthier approach to this issue as users can instruct their system as to which sound to play, and ideally the whole system would use that sound (kind of like macOS).

- Visual flashes: I think we can all agree that rapid flashes à la Konsole or Alacritty looks bad and hurts one's eyes (potentially life-threatening for people who suffer from photosensitive epilepsy). An alternative would be animated, smoother fading like what Kitty does, but that would require a rewrite of our graphics pipeline to support smooth animations without incurring large performance costs — a goal perhaps for 1.2, but it would take considerable effort to implement.

- System notifications/icon badges: Basically free.

- Focus requests/attention requests/jumping-up-and-down-on-the-dock: Also basically free. (I would consider making this enabled by default since it's a relatively unobtrusive way of grabbing the user's attention. It's also fairly standard on both Linux and macOS desktops.)

- Calling custom scripts: I am really against this. Running a script potentially hundreds of times a second when a program dumps out ^G like a flood sounds like a recipe for disaster. Ghostty is a terminal emulator, not a script runner.

**Reply by tristan957** [reply]
Here is what Ptyxis does, and actually I quite like the UX.


https://github.com/user-attachments/assets/7cf928e9-9fc7-479f-b919-0a5b5acac16e

**Reply by tristan957** [reply]
We would need to define the UX for a split. Maybe something like tmux focused splits? Highlight the split separators containing the surface?

**Reply by pluiedev** [reply]
Ooh, that's quite elegant. *Far* better idea of a "visual bell" than just flashing the screen a couple of times like a flashbang. I feel like maybe highlighting the split via a flashing border is the way to go here — that would also address something like #5330 where we similarly need to associate a surface event to the entire window.

**Reply by jcollie** [reply]
Is that using "adw_tab_page_set_needs_attention"? That would be quite easy to implement for us. I think for a split at minimum you'd want that split to become the focused split on that tab if a `BEL` is received. Could maybe add a CSS border around the split that fades after a few seconds?

**Reply by jcollie** [reply]
FYI, I updated the #5326 so that we don't need to include an audio theme in Ghostty. It implements two options `system` and `audio` - `system` uses the aforementioned `gdk_system_beep` to play the system beep, and `audio` now only plays a user-supplied audio file. Both of these options should be easily implementable on macOS.

**Reply by tristan957** [reply]
> Is that using "adw_tab_page_set_needs_attention"? 

Yes it is!

**Reply by dbast** [reply]
maybe beyond the scope of this discussion, but there are more "signals" to potentially react to than only special characters (like the bell) being detected ... like a user runs a specific command, e.g. `git push` triggering a specific action of the terminal  ... for things like https://github.com/bomanimc/gitrocket :)

**Reply by 00-kat** [reply]
That is probably out of the scope of this discussion, but do keep in mind that it may be possible to configure some tools to print out a bell character; for example, I had a in bell character in my sudo prompt so that I get notified whenever something tries to run sudo.

**Comment by jcollie** [comment, marked as answer]
#7087 fixes this for GTK. macOS support is a TODO.

**Reply by jcollie** [reply]
#7099 for tracking BELL support on macOS.
